src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs

```diff
@@ -872,23 +884,59 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
             public override void Initialize(AnalysisContext context)
             {
                 context.RegisterCompilationAction(compilationContext =>
-                    compilationContext.ReportDiagnostic(CodeAnalysis.Diagnostic.Create(UnsupportedDescriptor, Location.None)));
+                {
+                    try
+                    {
+                        ThrownException = null;
+                        compilationContext.ReportDiagnostic(CodeAnalysis.Diagnostic.Create(UnsupportedDescriptor, Location.None));
+                    }
+                    catch (Exception e)
+                    {
+                        ThrownException = e;
+                        throw;
+                    }
+                });
             }
         }
 
         [Fact, WorkItem(4376, "https://github.com/dotnet/roslyn/issues/4376")]
         public void TestReportingDiagnosticWithInvalidId()
         {
             string source = @"";
+            CSharpCompilation compilation = CreateCompilationWithMscorlib45(source);
             var analyzers = new DiagnosticAnalyzer[] { new AnalyzerWithInvalidDiagnosticId() };
             string message = new ArgumentException(string.Format(CodeAnalysisResources.InvalidDiagnosticIdReported, AnalyzerWithInvalidDiagnosticId.Descriptor.Id), "diagnostic").Message;
+            Exception analyzerException = null;
+            IFormattable context = $@"{string.Format(CodeAnalysisResources.ExceptionContext, $@"Compilation: {compilation.AssemblyName}")}
 
-            CreateCompilationWithMscorlib45(source)
-                .VerifyDiagnostics()
-                .VerifyAnalyzerDiagnostics(analyzers, null, null, logAnalyzerExceptionAsDiagnostics: true,
-                     expected: Diagnostic("AD0001")
-                     .WithArguments("Microsoft.CodeAnalysis.CommonDiagnosticAnalyzers+AnalyzerWithInvalidDiagnosticId", "System.ArgumentException", message)
-                     .WithLocation(1, 1));
+{new LazyToString(() => analyzerException)}
+-----
+
+{string.Format(CodeAnalysisResources.DisableAnalyzerDiagnosticsMessage, "Invalid ID")}";
+
+            EventHandler<FirstChanceExceptionEventArgs> firstChanceException = (sender, e) =>
+            {
+                if (e.Exception is ArgumentException)
+                {
+                    analyzerException = e.Exception;
+                }
```

Comment:
Are other exceptions expected here?

Reply:
No, a specific exception is thrown inside the analyzer engine and we a reference to it for the message details.